Low power stabilization

mlyle opened the pull request:
Fixes #228
- [x] Basic infrastructure for low power stabilization
- [x] Corrections to timer math for safety -- this was actually OK (proper windowed unsigned int subtraction)
- [x] Negative-side clip management implementation -- right now control authority is limited in that case
- ~~Need to give thought to integrator wind-up considerations in clipping scenarios~~ [issueopened]
- ~~Consider adding default actuator curve fit value = 0.85 or 0.9~~ [issueopened]
- ~~Consider adding actuator curve fit to UI~~ [issueopened]
- ~~Consider adding lowpower stab settings to UI~~ [issueopened]

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/d-ronin/dronin/676)

<!-- Reviewable:end -->

mlyle committed:
actuator: refactor to simplify "air mode" support

mlyle committed:
uavobjects/actuatorsettings: add low power stab values

mlyle committed:
flight/actuator: introduce low power stabilization mode

Still needs some work: cleanup of timer math, and use of the new
LowPowerStabilizationMaxPowerAdd parameter.

mlyle committed:
flight/actuator: clip neg throttle at 0 when low-power-stab

mlyle commented:
This is still very early; sketched out on a plane.  Unflown, and needs further work.

mlyle commented:
Setup requirements:

Channel neutral must be set with enough margin to ensure motor is guaranteed to spin well there.
Actuator curve fit is probably mandatory-- if value is unknown and the effects of actuator curve fit are undesirable, suggest a value of 0.9 to start.

mlyle committed:
flight/actuator: handle negative side clipping correctly

Also add 24 bytes to stacks to compensate for additional 8 bytes used out
of paranoia.

mlyle committed:
actuatorsettings: reconsider low-power-stab defaults

Also, update descriptions.

tracernz commented:
Should the curve fit parameter be added to actuator/output tab or stabilization tab? I think you could make a good case for either but more so for the actuator tab.

fujin commented:
jenkins, artifacts please

tracernz commented:
jenkins, build artifacts please

mlyle commented:
Jenkins couldn't talk to github-- socket error -- at completion to update artifact build status :/

mluessi commented on flight/Modules/Actuator/actuator.c:
```diff
@@ -295,10 +294,30 @@ static void actuator_task(void* parameters)
+			// Could consider stabilizing on a positive arming edge,
+			// but this seems problematic.
+		} else if (last_pos_throttle_time) {
+			if ((this_systime - last_pos_throttle_time) >
```
s/>/< ;)

mluessi replied:
Also, above we have `bool stabilize_now = throttle_source >= 0.0f;`, so this check gets bypassed whenever the stick is moved, no?

EDIT: Ignore this..

mlyle replied:
ah.  yes, that is a reversed test ;)

mluessi commented on flight/Modules/Actuator/actuator.c:
```diff
@@ -312,6 +331,8 @@ static void actuator_task(void* parameters)
 		float min_chan = INFINITY;
 		float max_chan = -INFINITY;
+		float neg_clip = 0;
+		int num_motors = 0;
```
I would call it `num_neg_clip` or something

mluessi commented on flight/Modules/Actuator/actuator.c:
```diff
@@ -343,29 +370,53 @@ static void actuator_task(void* parameters)
+			/* Low-side clip management-- how much power are we
+			 * willing to add??? */
+
+			neg_clip /= num_motors;
```
I wonder if using the mean is the right thing to do.. wouldn't using the maximum neg clip be better?

mlyle replied:
it's not num_neg_clip above.  It's the cumulative amount of clip.  What we do here is index it, and figure out how much the throttle has already been increased by clipping.

If we have 4 motors, and two are commanded to -.10 and two to .30.... the user has asked for 10% throttle, and we have sped up two motors 20% and slowed down two motors 20%.

But the actual throttle command achieved by this (assuming linearized actuators) is (0 + 0 + .30 + .30) / 4 = 15%.  So the throttle has already been increased by 5% by clipping.

If maximum throttle advance is 0%, the math below this point isn't quite right-- it's a bit of a fudge-- to lower each motor by 5% still doesn't quite hit the user throttle command-- it takes us to 12.5% in that case.  But really getting that right requires iteratively decreasing the throttle and watching how many motors are in clip, and it's considerably better than what we would do before.

Hm, maybe that can be further improved a little though.  Will think about it.

mlyle committed:
flight/actuator: fix reversed test in clipmgmt

fujin commented:
```
18:25:32 <fujin> sickness~~~~~ L.P.S works !
18:25:45 <fujin> using 15% lowpoweradd and 4secs lowpowertime
18:25:52 <fujin> probably will go up to 20% lowpoweradd
18:25:54 <fujin> maybe even more
18:26:01 <fujin> Need to go out to a larger area to test more
```

I will do more testing in a larger field with significant hang-time maneuvers. I just ran some quick laps around my house and tested some full power chops. Stability is still observed. Need more power.

flight vid: https://youtu.be/A6LyYEn2h5I

tracernz commented:
jenkins, build artifacts please

dustin commented:
Did some garage testing.  Pretty happy with how it was working.
http://gfycat.com/ImpoliteBiodegradableBarasingha

On Tue, Feb 23, 2016 at 9:10 PM Michael Corcoran notifications@github.com
wrote:

> jenkins, build artifacts please
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/d-ronin/dRonin/pull/676#issuecomment-188074071.

dustin commented:
This is something unexpected/undesirable that happened, though...  I cut
throttle and the engines cut out altogether.  I'm not sure if it crashed or
what, but it just stopped and fell over.  Only happened this one time.
There's enough white noise in the room (and the camera was in a waterproof
mount) that I couldn't make out whether it completely disarmed or not.  In
any case, had the normal amount of control for zero throttle and just
tumbled down.

http://gfycat.com/TidyIdolizedCurassow

On Tue, Feb 23, 2016 at 9:17 PM Dustin Sallings dustin@spy.net wrote:

> Did some garage testing.  Pretty happy with how it was working.
> http://gfycat.com/ImpoliteBiodegradableBarasingha
> 
> On Tue, Feb 23, 2016 at 9:10 PM Michael Corcoran notifications@github.com
> wrote:
> 
> > jenkins, build artifacts please
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/d-ronin/dRonin/pull/676#issuecomment-188074071.

tracernz commented:
Artifacts links for testers (Jenkins -> GitHub API currently broken)
- [windows](http://jar.lyle.org:8080/job/dronin/179/Nodes=winx86/artifact/build/artifacts/)
- [linux-x86_64](http://jar.lyle.org:8080/job/dronin/179/Nodes=master/artifact/build/artifacts/)
- [macosx](http://jar.lyle.org:8080/job/dronin/179/Nodes=osx/artifact/build/artifacts/)

fujin commented:
https://www.youtube.com/watch?v=d5NbP4PHoss

https://gfycat.com/MeekAnxiousBarasingha

https://gfycat.com/MammothRadiantDungbeetle

dustin commented:
^ dRonin: the movie.

On Tue, Feb 23, 2016 at 11:11 PM AJ Christensen notifications@github.com
wrote:

> https://www.youtube.com/watch?v=d5NbP4PHoss
> 
> https://gfycat.com/MeekAnxiousBarasingha
> 
> https://gfycat.com/MammothRadiantDungbeetle
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/d-ronin/dRonin/pull/676#issuecomment-188111520.

mlyle commented:
Going to open up issues for the remaining items, then submit this for merge.

mlyle commented:
P.S. @fujin thanks for being brave/crazy and trying this early :)

fujin commented:
@mlyle sure thing man :) happy to help.

fujin commented on flight/Modules/Actuator/actuator.c:
```diff
@@ -295,10 +294,30 @@ static void actuator_task(void* parameters)
+
+			// Could consider stabilizing on a positive arming edge,
+			// but this seems problematic.
+		} else if (last_pos_throttle_time) {
```
one or both of these if/elses need to check that the quad is actually armed first I think. the Brain OSD is generating `ACTUATOR` alarm warning while disarmed and with throttle low. I expect it is trying to LPS regardless of Armed state. @mlyle

mlyle replied:
I believe that will happen without this changeset, too.  In either case there's clip.  Perhaps open a bug?  I need to improve the actuator alarms in general before release because they're too chatty.

mlyle committed:
flight: failsafe: adjust behavior for hangtime

Now that we can stabilize with low throttle, failsafe is a little
improved for multirotor.  But it's essential we fall straight down for
everything except fixed wing.

mlyle commented:
Added a failsafe change to make failsafes + hangtime safer.  Discovered in code review.

mlyle commented:
stick-disarming can be kinda scary if LPS is still engaged.

mlyle commented:
LPS is amusing.

pug398 commented:
jenkins, build artifacts please

mlyle commented:
@pug398 -- recommendations for LPS
- switch arming
- duration < 4.5 seconds
- power add noticeably less than hover power

There's no UI for it in this one.. but if you've been flying kitchensink it's in there.  @fujin has flown it a lot, @dustin has flown it some, I've flown it a little.

It should be worth a tiny improvement in performance in normal flight, but the main thing is what it does during full-power-chops.

pug398 commented:
Yeah I started the build because I figured that was where the change was. I was just testing kitchensink overall for any interaction. I am not brave (good) enough to give LPS a workout :)

mlyle commented:
Added from IRC:  

@dustin http://gfycat.com/RemoteFrequentGonolek    http://gfycat.com/AbsolutePerfumedAsiansmallclawedotter

@fujin https://gfycat.com/ClumsyLeanKoodoo

dustin commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/676#-:-KBfLLftpR3NDGfQakCW:1417088721)_

<!-- Sent from Reviewable.io -->
